### PR TITLE
SIP-280 Upgradable PerpsV2 (+ SIP-279, SIP-281, SIP-285)

### PR DESCRIPTION
**Pull request by leomassazza**
Make current futures system upgradable, add a self-migrate mechanism so that users can migrate their positions from the older version to this one and add some hot/quick fixes to improve usability while waiting for the migration to V3.

Stage 0 is upgrading current Futures contracts to be upgradable and maintainable. The main architectural changes are:
- Remove all state from market contracts and move to a specific contract that can be migrated 
- Add a proxy in front of the FuturesMarket
- Split the main contract to make it fit in the max contract size

In order to achieve the 3 points above (mostly because of the need to split the FuturesMarkets), this PR:
- introduces a new set of contracts to keep the current version of Futures untouched. The new contracts names are based on **PerpsV2** prefix.
- introduces a new Proxy (ProxyFuturesV2.sol) based on Proxy.sol and taking the routing v3 concepts.
- introduces a shared state component (StateShared.sol) based on State.sol that adds the ability to use a state contract from more than one associatedContract.
- copies the classic Futures tests adapting them to the new architecture
- adds new tests to include coverage for the new stuff. 
- adds deployment scripts to cope with PerpsV2 architecture. 

### This PR: 

- [x] Create a copy of contracts for FuturesV2 (same functionality)
- [x] Update the architecture so that FuturesV2 is upgradable/maintainable 
  - [x]  Split FuturesV2Market into implementation and state
  - [x] Add a Proxy in front of the the implementation
  - [x] Update current tests to use new architecture
  - [x] Split contract to fit in size limits
    - [x] Improve proxy 
    - [x] Views
    - [x] Mutations (main txs)
    - [x] NextPriceOrders 
  - [x] Clean/Update comments
  - [x] Add tests for new contracts
    - [x] ProxyFuturesV2
    - [x] StateShared
    - [x] Security of sub contracts
  - [x] Update deploy tasks with new architecture
  - [x] Update/fix integration tests
  - [x] Remove PerpsV2 references and rename FuturesV2 to PerpsV2

### Tasks 
  - [x] Merge Stage 1 branches
  - [x] Update from Develop and unit fix tests
  - [ ] Apply Audit Fixes 
  - [x] Simplify Migration
  - [x] Fix missing tests
 
### Feature Branches Stage 1
- [x] Feature/Fix: Make Next Price Order time based (https://github.com/Synthetixio/synthetix/pull/1899)
- [x] Feature/Fix: Fix Maker/Taker for cross-sides orders (https://github.com/Synthetixio/synthetix/pull/1906)
- [x] Feature: Off-chain price feeds Delayed Orders (https://github.com/Synthetixio/synthetix/pull/1913)
- [x] Feature/Fix: Upgrade Funding Mechanism (https://github.com/Synthetixio/synthetix/pull/1914)
### Feature Branches Stage 2
- [x] Feature/Fix: Price Correction (https://github.com/Synthetixio/synthetix/pull/1921)
- [x] Feature/Fix: Price impact protection mechanism (https://github.com/Synthetixio/synthetix/pull/1936)
- [x] Feature/fix: Liquidations to consider premium (https://github.com/Synthetixio/synthetix/pull/1943)
### Feature Branches Optionals
- [ ] Feature/Fix: (optional) Intend to trade
- [ ] Feature/Fix: (optional) Margin with other synths

### CODE REVIEW

File: contracts/PerpsV2MarketProxyable.sol [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+        uint price,
+        uint fee
+    ) internal {
+        proxy._emit(abi.encode(id, account, liquidator, size, price, fee), 1, POSITIONLIQUIDATED_SIG, 0, 0, 0);
+    }
+
+    event FundingRecomputed(int funding, uint index, uint timestamp);
+    bytes32 internal constant FUNDINGRECOMPUTED_SIG = keccak256("FundingRecomputed(int256,uint256,uint256)");
+
+    function emitPositionLiquidated(
```
Comment (davidvuong) [inline contracts/PerpsV2MarketProxyable.sol:303]: @leomassazza Caught this while implementing new funding. I don't think this looks right. It should be `emitFundingRecomputed`.

Comment (leomassazza) [reply contracts/PerpsV2MarketProxyable.sol]: fixed

**Comment by davidvuong** [comment]
https://github.com/Synthetixio/synthetix/runs/9300428570 This has been running for a while now 

> Started 2d 7h 21m 54s ago

😱

**Comment by davidvuong** [comment]
I don't think the DelayedOrder changes in https://github.com/Synthetixio/synthetix/pull/1899 were merged correctly. There are still instances where NextPrice is still present. Same with the new funding model https://github.com/Synthetixio/synthetix/pull/1914. Changes to incorporate velocity don't seem to be present here.

File: contracts/PerpsV2ExchangeRate.sol [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        uint fee = offchainOracle().getUpdateFee(priceUpdateData.length);
+        require(msg.value >= fee, "Not enough eth for paying the fee");
+
+        // Update the price data (and pay the fee)
+        offchainOracle().updatePriceFeeds.value(fee)(priceUpdateData);
+
+        if (msg.value - fee > 0) {
+            // Need to refund caller. Try to return unused value, or revert if failed
+            // solhint-disable-next-line  avoid-low-level-calls
+            (bool success, ) = sender.call.value(msg.value - fee)("");
```
Comment (kylriley) [inline contracts/PerpsV2ExchangeRate.sol:67]: Just need to note that this allows users to call (w/ eth value) the fallback functions of arbitrary contracts, including  internal Synthetix contracts. Very unlikely to be an issue, but worth being aware of.

**Review by barrasso — APPROVED** [review]

**Pull request merged**